airflow/contrib/operators/gcp_function_operator.py

```diff
@@ -0,0 +1,571 @@
+
+    def _create_new_function(self):
+        self._hook.create_new_function(self.location, self.body)
+
+    def _update_function(self):
+        self._hook.update_function(self.body['name'], self.body, self.body.keys())
+
+    def _check_if_function_exists(self):
+        return self.body['name'] in map(
+            lambda x: x["name"], self._hook.list_functions(self.location)
+        )
+
+    def _upload_source_code(self):
+        return self._hook.upload_function_zip(parent=self.location,
+                                              zip_path=self.zip_path)
+
+    def execute(self, context):
+        if self.source_code_validation.should_upload_function():
+            self.body[SOURCE_UPLOAD_URL] = self._upload_source_code()
+        self._validate_all_body_fields()
+        if not self._check_if_function_exists():
+            self._create_new_function()
+        else:
+            self._update_function()
+
+
+class SourceCodeValidation:
+    """
+    Validates Source Code parameters.
+
+    Responsible for validating if we have enough information about the source code of
+    the GCF function. The source code to be used can be specified using different options:
+    * non-empty sourceUploadUrl in case the code was just uploaded using
+      "generateUploadUrl" mechanism
+    * empty sourceUploadUrl and path to zip file (then generateUploadUrl will
+      be used to upload the zip file and sourceUploadUrl will be set appropriately
+    * archiveUrl that points to valid Google Storage link (starting with gs://)
+      where the zip file is uploaded.
+
+    :param body: Body passed to the create/update method calls.
+    :type body: dict
+    :param zip_path: path to the zip file containing source code. If it is
+        set, then sourceUploadUrl should not be specified in the body (or it should
+        be empty), then the zip file will be uploaded using upload URL generated
+        via generateUploadUrl from cloud functions API
+    :type body: str
+
+    """
+    upload_function = None
+
+    def __init__(self, body, zip_path):
+        self.body = body
+        self.zip_path = zip_path
+
+    @staticmethod
+    def _is_present_and_empty(dictionary, field):
+        return field in dictionary and not dictionary[field]
+
+    def _is_upload_url_and_no_zip_path(self):
+        if self._is_present_and_empty(self.body, SOURCE_UPLOAD_URL):
+            if not self.zip_path:
+                raise AirflowException(
+                    "If parameter '{}' is empty in 'body' then argument '{}' needed.".
+                    format(SOURCE_UPLOAD_URL, ZIP_PATH))
+
+    def _is_upload_url_and_zip_path(self):
+        if SOURCE_UPLOAD_URL in self.body and self.zip_path:
+            if not self.body[SOURCE_UPLOAD_URL]:
+                self.upload_function = True
+            else:
+                raise AirflowException("Only one of '{}' in body or '{}' argument "
+                                       "possible. Found both."
```

Comment:
Change 'possible' to 'allowed'.

Reply:
Right :).

Reply:
Fixed in b9027de5b2b2f14d7b3afecd3fa1b2a692a9b914